src/app/ReadHandler.cpp

```diff
@@ -446,5 +513,100 @@ void ReadHandler::MoveToNextScheduledDirtyPriority()
 
     mCurrentPriority = PriorityLevel::Invalid;
 }
+
+CHIP_ERROR ReadHandler::SendSubscribeResponse()
+{
+    SubscribeResponse::Builder response;
+    System::PacketBufferTLVWriter writer;
+    System::PacketBufferHandle packet = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
+    VerifyOrReturnLogError(!packet.IsNull(), CHIP_ERROR_NO_MEMORY);
+
+    writer.Init(std::move(packet));
+    ReturnLogErrorOnFailure(response.Init(&writer));
+    response.SubscriptionId(mSubscriptionId).FinalSyncIntervalSeconds(mMaxIntervalSeconds).EndOfSubscribeResponse();
+    ReturnLogErrorOnFailure(response.GetError());
+
+    ReturnLogErrorOnFailure(writer.Finalize(&packet));
+    VerifyOrReturnLogError(mpExchangeCtx != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    MoveToState(HandlerState::SubscribingResponding);
+    ReturnLogErrorOnFailure(mpExchangeCtx->SendMessage(Protocols::InteractionModel::MsgType::SubscribeResponse, std::move(packet),
+                                                       Messaging::SendFlags(Messaging::SendMessageFlags::kExpectResponse)));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR ReadHandler::ProcessSubscribeRequest(System::PacketBufferHandle && aPayload)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    System::PacketBufferTLVReader reader;
+    SubscribeRequest::Parser subscribeRequestParser;
+    EventPathList::Parser eventPathListParser;
+    AttributePathList::Parser attributePathListParser;
+
+    reader.Init(std::move(aPayload));
+
+    ReturnLogErrorOnFailure(reader.Next());
+    ReturnLogErrorOnFailure(subscribeRequestParser.Init(reader));
+#if CHIP_CONFIG_IM_ENABLE_SCHEMA_CHECK
+    ReturnLogErrorOnFailure(subscribeRequestParser.CheckSchemaValidity());
+#endif
+
+    err = subscribeRequestParser.GetAttributePathList(&attributePathListParser);
+    if (err == CHIP_END_OF_TLV)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    else if (err == CHIP_NO_ERROR)
+    {
+        ReturnLogErrorOnFailure(ProcessAttributePathList(attributePathListParser));
+    }
+    ReturnLogErrorOnFailure(err);
+
+    err = subscribeRequestParser.GetEventPathList(&eventPathListParser);
+    if (err == CHIP_END_OF_TLV)
+    {
+        err = CHIP_NO_ERROR;
+    }
+    else if (err == CHIP_NO_ERROR)
+    {
+        ReturnLogErrorOnFailure(ProcessEventPathList(eventPathListParser));
+    }
+    ReturnLogErrorOnFailure(err);
+
+    ReturnLogErrorOnFailure(subscribeRequestParser.GetMinIntervalSeconds(&mMinIntervalSeconds));
+    ReturnLogErrorOnFailure(subscribeRequestParser.GetMaxIntervalSeconds(&mMaxIntervalSeconds));
+
+    // TODO: Use GetSecureRandomData to generate subscription id
+    // err = Platform::Security::GetSecureRandomData((uint8_t *) &mSubscriptionId, sizeof(mSubscriptionId));
```

Comment:
can we do this now? rand() is really unsafe.

Reply:
RandUtils should be ok - we will have a separate task to make randutils use the DRNG that spec decides (i.e. it will be secure)

Reply:
sure, thanks